src/Uno.UI/UI/Xaml/Internal/InputManager.Keyboard.skia.cs

```diff
@@ -44,60 +44,38 @@ public void Init(object host)
 				CoreWindow.GetForCurrentThreadSafe()?.SetKeyboardInputSource(_source);
 			}
 
-			_source.KeyDown += (s, e) => InitiateKeyDownBubblingFlow(e);
-			_source.KeyUp += (s, e) => InitiateKeyUpBubblingFlow(e);
+			_source.KeyDown += (_, e) => OnKey(e, true);
+			_source.KeyUp += (_, e) => OnKey(e, false);
 		}
 
-		private void InitiateKeyDownBubblingFlow(KeyEventArgs args)
+		private void OnKey(KeyEventArgs args, bool down)
 		{
-			var originalSource = FocusManager.GetFocusedElement(_inputManager.ContentRoot.XamlRoot) as UIElement ?? _inputManager.ContentRoot.VisualTree.RootElement;
+			var originalSource1 = FocusManager.GetFocusedElement(_inputManager.ContentRoot.XamlRoot) as UIElement ?? _inputManager.ContentRoot.VisualTree.RootElement;
 
-			if (originalSource is null)
-			{
-				return;
-			}
-
-			var krea = new KeyRoutedEventArgs(originalSource, args.VirtualKey, args.KeyboardModifiers, args.KeyStatus, args.UnicodeKey)
+			var args1 = new KeyRoutedEventArgs(originalSource1, args.VirtualKey, args.KeyboardModifiers, args.KeyStatus, args.UnicodeKey)
 			{
 				CanBubbleNatively = false
 			};
-			originalSource.RaiseEvent(UIElement.KeyDownEvent, krea);
-			args.Handled = krea.Handled;
 
-			if (this.Log().IsEnabled(LogLevel.Trace))
-			{
-				this.Log().Trace(
-					$"CoreWindow_KeyDown(vk: {args.VirtualKey}, " +
-					$"IsExtendedKey: {args.KeyStatus.IsExtendedKey}, " +
-					$"IsKeyReleased: {args.KeyStatus.IsKeyReleased}, " +
-					$"IsMenuKeyDown: {args.KeyStatus.IsMenuKeyDown}, " +
-					$"RepeatCount: {args.KeyStatus.RepeatCount}, " +
-					$"ScanCode: {args.KeyStatus.ScanCode})"
-				);
-			}
-		}
+			originalSource1.RaiseTunnelingEvent(down ? UIElement.PreviewKeyDownEvent : UIElement.PreviewKeyUpEvent, args1);
 
-		private void InitiateKeyUpBubblingFlow(KeyEventArgs args)
-		{
-			var originalSource = FocusManager.GetFocusedElement(_inputManager.ContentRoot.XamlRoot) as UIElement ?? _inputManager.ContentRoot.VisualTree.RootElement;
+			// On WinUI, if the focus changes during PreviewKey<Down|Up>, the Key<Up|Down> event bubbles from the new focused element.
+			var originalSource2 = FocusManager.GetFocusedElement(_inputManager.ContentRoot.XamlRoot) as UIElement ?? _inputManager.ContentRoot.VisualTree.RootElement;
 
-			if (originalSource is null)
+			var args2 = new KeyRoutedEventArgs(originalSource2, args.VirtualKey, args.KeyboardModifiers, args.KeyStatus, args.UnicodeKey)
```

Comment:
As we are not re-using neither `originalSource` nor `args` I would just update the variable ... it would insignificantly reduce memory pressure.

BTW, I would use `args` and `routedArgs` ... numbering variable is usually not really meaningful.

Reply:
We are not reusing it, but WinUI doesn't use the same args object twice. If it's okay to not match WinUI here, then I agree, it's totally unnecessary.

Reply:
No strong opinion on this, do what you think is the best :)